libass/ass_render.c

```diff
@@ -1392,15 +1405,18 @@ get_bitmap_glyph(ASS_Renderer *render_priv, GlyphInfo *info,
     }
 
     key.outline = ass_cache_get(render_priv->cache.outline_cache, &ol_key, render_priv);
-    if (!key.outline || !key.outline->valid || !quantize_transform(m, pos_o, &key)) {
+    if (!key.outline || !key.outline->valid ||
+            !quantize_transform(m, pos_o, offset, false, &key)) {
         ass_cache_dec_ref(key.outline);
         return;
     }
     info->bm_o = ass_cache_get(render_priv->cache.bitmap_cache, &key, render_priv);
     if (!info->bm_o || !info->bm_o->buffer) {
         ass_cache_dec_ref(info->bm_o);
         info->bm_o = NULL;
-    }
+        *pos_o = *pos;
+    } else if (!info->bm)
+        *pos = *pos_o;
```

Comment:
I'm setting position of NULL bitmap to position of alive bitmap. `pos` and `pos_o` goes to composite cache key so it should have well defined values in case of empty bitmaps.

Reply:
Yes, I don't care about position of NULL bitmap and don't want to pollute cache key with it. In case of both NULL bitmaps they'll be skipped altogether so exact value of `pos` and `pos_o` don't matter.

Reply:
OK. Don’t you think it would be clearer to set them to 0 immediately after each `ass_cache_get` call if the call fails? (With or without a comment mentioning cache efficiency. Ideally with, but e. g. for shadow.x/y in the fix-shadows commit you don’t have a comment IIRC; but when it’s so obviously special-cased, it’s not a huge deal.)

Reply:
You also want to reset `*pos_o` in case `quantize_transform(m, pos_o, &key)` fails several lines above. `quantize_transform` can fail after writing the position.

Reply:
> Don’t you think it would be clearer to set them to 0 immediately after each `ass_cache_get` call if the call fails?

`pos` and `pos_o` will be modified later (composite bitmap origin calculation and moving it to zero) so zero value will be offset by some random value and pollute cache.

> `quantize_transform` can fail after writing the position.

I've fixed that in "renderer: synchronize glyph motion to help composite cache". I'll try to move that fix earlier.

Reply:
> I'll try to move that fix earlier.

Done.